### ISSUE TITLE
Some questions about the code

### ISSUE THREAD

**Issue by 20chase** [pat-coady/trpo#5]
Hi pat,

Good job! This code have a excellent performance in continue space env. But I have some questions about the code. 

1. I can't understand the meaning of logvar_speed, could you explain that why it can make faster updates or give me some paper or blog something like these to understand it?

2. Why value function need use the 2 batches to update? which means every batch need to be updated 2 time.

3. Why checking variance before and after update value function can diagnose over-fitting?

I am looking forward your reply. : )

**Comment by pat-coady** [comment]
@chase20

Sorry for the slow reply - it has been a very busy week.

1) The learning rate is partially determined by the depth and size of the NN. The 
derivative of policy loss with respect to each weight is
calculated independently. So several weights will tend to constructively
combine and push the policy too far and cause the update to overshoot. I'm
sure you've seen the cartoons where the update jumps across the convexity
and makes the loss worse. So you lower the learning rate to where you get
good improvement in training loss with each gradient update.

Unfortunately the learning rate the policy NN needs is much slower than
the learning rate the variance variables can handle. These the variance tensor
is just a trainable variable with no NN. So by making the variance the sum 
of 1x"logvar_speed" variables, this increases the gradient update to variance by a factor of
"logvar_speed". This was a bit of a hack, a cleaner approach would have
been to just multiply the grads of variance variable by log_speed.

This came about because I noticed that the variance was training 
much more slowly than I expected. It
was an almost perfect linear ramp, which indicated to me that it was
limited by the learning rate. 

This change made a big difference in learning performance - maybe the single 
biggest difference of improvements I made.

2) Just to smooth things out. Sometimes the policy gets worse after an
update, you don't want the Q function to chase this around. So, in some
ways, this is similar to a replay buffer in A3C if you're familiar with
that.

3) Somewhat related to item 2. With TRPO we are limiting the policy change
between training iterations. So, the Q values of the policy shouldn't
change too much either. So if the old Q function only explains 50% of the variance
after an update, but the newly fitted Q function explains 95% of the
variance, then you've over-fit the last batch of episodes and your Q
function isn't likely to extrapolate well when calculating next batch of
advantages.

A bit long-winded, but I hope this helps.

Pat
